native/dlaqr4.go

```diff
@@ -0,0 +1,467 @@
+	nwr = max(2, nwr)
+	nwr = min(ihi-ilo+1, min((n-1)/3, nwr))
+
+	// nsr is the recommended number of simultaneous shifts. n is greater
+	// than 11, so there is enough subdiagonal workspace for nsr to be even
+	// and greater than or equal to two as required.
+	nsr := impl.Ilaenv(15, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	nsr = min(nsr, min((n+6)/9, ihi-ilo))
+	nsr = max(2, nsr-nsr%2)
+
+	// Workspace query call to Dlaqr2.
+	impl.Dlaqr2(wantt, wantz, n, ilo, ihi, nwr+1, nil, 0, iloz, ihiz, nil, 0,
+		nil, nil, nil, 0, n, nil, 0, n, nil, 0, work, -1)
+	// Optimal workspace is max(Dlaqr5, Dlaqr2).
+	lwkopt := max(3*nsr/2, int(work[0]))
+	// Quick return in case of workspace query.
+	if lwork == -1 {
+		work[0] = float64(lwkopt)
+		return 0
+	}
+
+	// Dlahqr/Dlaqr4 crossover point.
+	nmin := impl.Ilaenv(12, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	nmin = max(ntiny, nmin)
+
+	// Nibble determines when to skip a multi-shift QR sweep (Dlaqr5).
+	nibble := impl.Ilaenv(14, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	nibble = max(0, nibble)
+
+	// Computation mode of far-from-diagonal orthogonal updates in Dlaqr5.
+	kacc22 := impl.Ilaenv(16, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	kacc22 = max(0, min(kacc22, 2))
+
+	// nwmax is the largest possible deflation window for which there is
+	// sufficient workspace.
+	nwmax := min((n-1)/3, lwork/2)
+
+	// nsmax is the largest number of simultaneous shifts for which there is
+	// sufficient workspace.
+	nsmax := min((n+6)/9, 2*lwork/3)
+	nsmax -= nsmax % 2
+
+	var (
+		ndec int     // Deflation window size decrement.
+		ndfl = 1     // Number of iteration since last deflation.
+		nw   = nwmax // Start with maximum deflation window size.
+
+		itmax = max(30, 2*kexsh) * max(10, (ihi-ilo+1))
+		it    = 0
+	)
+	// Main loop.
+	for kbot := ihi; kbot >= ilo; {
+		if it == itmax {
+			unconverged = kbot + 1
+			break
+		}
+		it++
+
+		// Locate active block.
+		ktop := ilo
+		for k := kbot; k >= ilo+1; k-- {
+			if h[k*ldh+k-1] == 0 {
+				ktop = k
+				break
+			}
+		}
+
+		// Select deflation window size nw.
+		//
+		// Typical Case:
+		//  If possible and advisable, nibble the entire active block.
+		//  If not, use size min(nwr,nwmax) or min(nwr+1,nwmax)
+		//  depending upon which has the smaller corresponding
+		//  subdiagonal entry (a heuristic).
+		//
+		// Exceptional Case:
+		//  If there have been no deflations in kexnw or more
+		//  iterations, then vary the deflation window size. At first,
+		//  because, larger windows are, in general, more powerful than
+		//  smaller ones, rapidly increase the window to the maximum
+		//  possible. Then, gradually reduce the window size.
+		nh := kbot - ktop + 1
+		nwupbd := min(nh, nwmax)
+		if ndfl < kexnw {
+			nw = min(nwupbd, nwr)
+		} else {
+			nw = min(nwupbd, 2*nw)
+		}
+		if nw < nwmax {
+			if nw >= nh-1 {
+				nw = nh
+			} else {
+				kwtop := kbot - nw + 1
+				if math.Abs(h[kwtop*ldh+kwtop-1]) > math.Abs(h[(kwtop-1)*ldh+kwtop-2]) {
```

Comment:
Lines 496-497 in the [Fortran](http://www.netlib.org/lapack/explore-html/db/d3b/dlaqr4_8f_source.html) are like something out of [this](http://mindprod.com/jgloss/unmain.html) at an expert level.